File: day9.py
import sys
from itertools import combinations, pairwise


points = [] 

for l in sys.stdin:
    l = l.strip()
    if not l:
        continue
    x, y = l.split(',')
    x = int(x)
    y = int(y)
    points.append((x, y))
    
res1 = 0

for p1, p2 in combinations(points, 2):
    area = (abs(p1[0] - p2[0]) + 1)  * (abs(p1[1] - p2[1]) + 1)
    res1 = max(res1, area)

print("Part 1:", res1)


points.append(points[0])

def crosses(tl, br):
    
    for p3, p4 in pairwise(points):
        
        tl_2 = (min(p3[0], p4[0]), min(p3[1], p4[1]))
        br_2 = (max(p3[0], p4[0]), max(p3[1], p4[1]))
        
        h_2 = tl_2[1] == br_2[1]
        
        if h_2:
            
            if tl_2[1] <= tl[1] or tl_2[1] >= br[1]:
                continue
            
            if br_2[0] > tl[0] and tl_2[0] < br[0]: 
                return True
        else:
            if tl_2[0] <= tl[0] or tl_2[0] >= br[0]:
                continue
            if br_2[1] > tl[1] and tl_2[1] < br[1]: 
                return True

    return False

res_2 = 0
   
idx = 0  
n_idx = 1
 
for p1, p2 in combinations(points, 2):
    idx += 1
    if idx == n_idx:
        n_idx *= 2
        print(f"Progress: {idx} / {len(points)*(len(points)-1)//2}")
    
    x1, y1 = p1
    x2, y2 = p2
    
    min_p = min(x1, x2), min(y1, y2)
    max_p = max(x1, x2), max(y1, y2)

    area = (max_p[0] - min_p[0] + 1) * (max_p[1] - min_p[1] + 1)
    
    if crosses(min_p, max_p):
        continue

    if res_2 < area:
        print(f"New max area: {area} from points {p1} and {p2}")
        res_2 = area
    
print("Part 2:", res_2)

File: day9.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>
#include <array>
#include <deque>
#include <
using namespace std;

std::vector<std::pair<int64_t, int64_t>> parsePairFile(const std::string& filename) {
    std::vector<std::pair<int64_t, int64_t>> pairs;
    std::ifstream file(filename);
    
    if (!file.is_open()) {
        std::cerr << "Error: Could not open file " << filename << std::endl;
        return pairs;
    }
    
    std::string line;
    while (std::getline(file, line)) {
        size_t start = line.find_first_not_of(" \t\n\r");
        size_t end = line.find_last_not_of(" \t\n\r");
        
        if (start == std::string::npos) continue; // Empty line
        
        line = line.substr(start, end - start + 1);
        
        size_t commaPos = line.find(',');
        if (commaPos != std::string::npos) {
            try {
                int64_t first = std::stoll(line.substr(0, commaPos));
                int64_t second = std::stoll(line.substr(commaPos + 1));
                pairs.push_back(std::make_pair(first, second));
            } catch (const std::exception& e) {
                std::cerr << "Error parsing line: " << line << std::endl;
            }
        }
    }
    
    file.close();
    return pairs;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <input_file>" << std::endl;
        return 1;
    }
    auto pairs = parsePairFile(argv[1]);



    int64_t res1 = 0;
    for (int i = 0; i + 1< pairs.size(); ++i) {
        for (int j = i + 1; j + 1 < pairs.size(); ++j) {
            const auto& p1 = pairs[i];
            const auto& p2 = pairs[j];

            pair<int64_t, int64_t> tl = {std::min(p1.first, p2.first), std::min(p1.second, p2.second)};
            pair<int64_t, int64_t> br = {std::max(p1.first, p2.first), std::max(p1.second, p2.second)};

            int64_t area = (br.first - tl.first + 1) * (br.second - tl.second + 1);
            res1 = max(res1, area); 
        }
    }

    std::cout << "Max area: " << res1 << std::endl;

    pair<int64_t, int64_t> br = {0ll, 0ll};
    for (auto& p : pairs) {
        p.first += 1; 
        p.second += 1;
        br = {max(br.first, p.first), max(br.second, p.second)};
    }

    br.first += 2;
    br.second += 2;

    vector<vector<uint8_t>> grid(br.first, vector<uint8_t>(br.second, 1));

    for(int i = 0; i < pairs.size(); ++i) {
        auto& p1 = pairs[i];
        auto& p2 = pairs[(i + 1) % pairs.size()];

        pair<int64_t, int64_t> tl2 = {std::min(p1.first, p2.first), std::min(p1.second, p2.second)};
        pair<int64_t, int64_t> br2 = {std::max(p1.first, p2.first), std::max(p1.second, p2.second)};

        for(int x = tl2.first; x <= br2.first; ++x) {
            for(int y = tl2.second; y <= br2.second; ++y) {
                grid[x][y] = 2;
            }
        }
    }

    std::deque<pair<int64_t, int64_t>> q;
    q.push_back({0,0});
    grid[0][0] = 0;
    int64_t count = 0;
    int64_t print_next = 1;

    while(q.size()) {

        ++count;
        if (count == print_next) {
            std::cout << "flooded " << count << " fields" << std::endl;
            print_next *= 2;
        }

        auto [x, y] = q.front();
        q.pop_front();

        array<pair<int64_t, int64_t>, 4> neigs = {{
                  {x + 1, y},
                    {x - 1, y},
                    {x, y + 1},
                    {x, y - 1}       
        }};

        for (const auto& neig : neigs) {
            auto [nx, ny] = neig;
            if (nx < 0 || ny < 0 || nx >= br.first || ny >= br.second) {
                continue;
            }

            if (grid[nx][ny] != 1) {
                continue;
            }

            grid[nx][ny] = 0;
            q.push_back({nx, ny});
        }
    }

    auto check_no_zero = [&](pair<int64_t, int64_t> p1, pair<int64_t, int64_t> p2) {
        for(int x = p1.first; x <= p2.first; ++x) {
            for(int y = p1.second; y <= p2.second; ++y) {
                if (grid[x][y] == 0) {
                    return false;
                }
            }
        }
        return true;
    };

    int64_t res2 = 0;

    count = 0;
    print_next = 1;
    for (int i = 0; i + 1< pairs.size(); ++i) {
        for (int j = i + 1; j + 1 < pairs.size(); ++j) {
            ++count;
            if (count == print_next) {
                std::cout << "Checked " << count << " pairs" << std::endl;
                print_next *= 2;
            }
            const auto& p1 = pairs[i];
            const auto& p2 = pairs[j];


            pair<int64_t, int64_t> tl = {std::min(p1.first, p2.first), std::min(p1.second, p2.second)};
            pair<int64_t, int64_t> br = {std::max(p1.first, p2.first), std::max(p1.second, p2.second)};

            if (!check_no_zero({tl.first, tl.second}, {br.first, tl.second})
                || !check_no_zero({tl.first, tl.second}, {tl.first, br.second})
                || !check_no_zero({br.first, tl.second}, {br.first, br.second})
                || !check_no_zero({tl.first, br.second}, {br.first, br.second})) {
                continue;
            }
            int64_t area = (br.first - tl.first + 1) * (br.second - tl.second + 1);
            res2 = max(res2, area); 
        }
    }

    cout << "Max area no flood: " << res2 << endl;
    
    return 0;
}